File: apps/mcp-context-engine/test/integration/fixtures/sample_code.py
#!/usr/bin/env python3
"""
Sample Python code for integration testing
This module demonstrates various Python constructs for testing search functionality
"""

import os
import sys
import json
import hashlib
from typing import List, Dict, Optional, Union
from dataclasses import dataclass
from datetime import datetime


@dataclass
class User:
    """User data class"""
    id: int
    name: str
    email: str
    created_at: datetime
    

class UserManager:
    """Manages user operations"""
    
    def __init__(self):
        self.users: List[User] = []
        self._user_cache: Dict[int, User] = {}
    
    def authenticate(self, username: str, password: str) -> Optional[User]:
        """Authenticate a user with username and password"""
        # Mock authentication - in real system would check database
        if username == "admin" and password == "secret123":
            return User(
                id=1,
                name="Administrator", 
                email="admin@example.com",
                created_at=datetime.now()
            )
        return None
    
    def add_user(self, user: User) -> bool:
        """Add a user to the system"""
        if self.get_user_by_email(user.email):
            return False  # User already exists
        
        self.users.append(user)
        self._user_cache[user.id] = user
        return True
    
    def get_user_by_email(self, email: str) -> Optional[User]:
        """Get user by email address"""
        for user in self.users:
            if user.email == email:
                return user
        return None
    
    def get_user_stats(self) -> Dict[str, Union[int, List[str]]]:
        """Get statistics about users"""
        domains = {}
        for user in self.users:
            domain = self._extract_domain(user.email)
            domains[domain] = domains.get(domain, 0) + 1
        
        return {
            "total_users": len(self.users),
            "domains": list(domains.keys()),
            "domain_counts": domains
        }
    
    def _extract_domain(self, email: str) -> str:
        """Extract domain from email address"""
        if "@" not in email:
            return "invalid"
        return email.split("@")[1]


def hash_password(password: str) -> str:
    """Hash a password using SHA256"""
    return hashlib.sha256(password.encode()).hexdigest()


def validate_email(email: str) -> bool:
    """Validate email format"""
    return "@" in email and "." in email.split("@")[1]


def load_config(config_path: str) -> Dict:
    """Load configuration from JSON file"""
    try:
        with open(config_path, 'r') as f:
            return json.load(f)
    except FileNotFoundError:
        print(f"Config file not found: {config_path}")
        return {}
    except json.JSONDecodeError:
        print(f"Invalid JSON in config file: {config_path}")
        return {}


def process_user_data(users_data: List[Dict]) -> List[User]:
    """Process raw user data into User objects"""
    users = []
    for data in users_data:
        try:
            user = User(
                id=data["id"],
                name=data["name"],
                email=data["email"],
                created_at=datetime.fromisoformat(data.get("created_at", datetime.now().isoformat()))
            )
            users.append(user)
        except (KeyError, ValueError) as e:
            print(f"Error processing user data: {e}")
            continue
    
    return users


if __name__ == "__main__":
    # Demo usage
    manager = UserManager()
    
    # Test authentication
    user = manager.authenticate("admin", "secret123")
    if user:
        print(f"Authentication successful for {user.name}")
    else:
        print("Authentication failed")
        sys.exit(1)
    
    # Add some test users
    test_users = [
        User(1, "Alice Johnson", "alice@example.com", datetime.now()),
        User(2, "Bob Smith", "bob@company.org", datetime.now()),
        User(3, "Charlie Brown", "charlie@example.com", datetime.now()),
    ]
    
    for user in test_users:
        if manager.add_user(user):
            print(f"Added user: {user.name}")
        else:
            print(f"Failed to add user: {user.name}")
    
    # Get statistics
    stats = manager.get_user_stats()
    print(f"User statistics: {json.dumps(stats, indent=2)}")

File: apps/mcp-context-engine/test/integration/fixtures/sample_code.js
/**
 * Sample JavaScript code for integration testing
 * Demonstrates various JavaScript patterns for search functionality testing
 */

import crypto from 'crypto';
import fs from 'fs/promises';
import path from 'path';

/**
 * User class representing a system user
 */
class User {
    constructor(id, name, email, createdAt = new Date()) {
        this.id = id;
        this.name = name;
        this.email = email;
        this.createdAt = createdAt;
    }

    /**
     * Get the domain from user's email
     */
    getDomain() {
        return this.email.split('@')[1] || 'unknown';
    }

    /**
     * Convert user to JSON representation
     */
    toJSON() {
        return {
            id: this.id,
            name: this.name,
            email: this.email,
            createdAt: this.createdAt.toISOString()
        };
    }
}

/**
 * UserManager handles user operations
 */
class UserManager {
    constructor() {
        this.users = new Map();
        this.emailIndex = new Map();
    }

    /**
     * Authenticate user with username and password
     */
    async authenticate(username, password) {
        // Mock authentication - would normally check database
        if (username === 'admin' && password === 'secret123') {
            return new User(1, 'Administrator', 'admin@example.com');
        }
        return null;
    }

    /**
     * Add a new user to the system
     */
    addUser(user) {
        if (this.emailIndex.has(user.email)) {
            return false; // User already exists
        }

        this.users.set(user.id, user);
        this.emailIndex.set(user.email, user.id);
        return true;
    }

    /**
     * Get user by email address
     */
    getUserByEmail(email) {
        const userId = this.emailIndex.get(email);
        return userId ? this.users.get(userId) : null;
    }

    /**
     * Get user by ID
     */
    getUserById(id) {
        return this.users.get(id) || null;
    }

    /**
     * Get all users as array
     */
    getAllUsers() {
        return Array.from(this.users.values());
    }

    /**
     * Get user statistics
     */
    getUserStats() {
        const domains = new Map();
        const allUsers = this.getAllUsers();

        for (const user of allUsers) {
            const domain = user.getDomain();
            domains.set(domain, (domains.get(domain) || 0) + 1);
        }

        return {
            totalUsers: allUsers.length,
            domains: Array.from(domains.keys()),
            domainCounts: Object.fromEntries(domains)
        };
    }

    /**
     * Search users by name pattern
     */
    searchUsers(pattern) {
        const regex = new RegExp(pattern, 'i');
        return this.getAllUsers().filter(user => 
            regex.test(user.name) || regex.test(user.email)
        );
    }
}

/**
 * Hash password using SHA256
 */
function hashPassword(password) {
    return crypto.createHash('sha256').update(password).digest('hex');
}

/**
 * Validate email format
 */
function validateEmail(email) {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
}

/**
 * Load configuration from JSON file
 */
async function loadConfig(configPath) {
    try {
        const data = await fs.readFile(configPath, 'utf8');
        return JSON.parse(data);
    } catch (error) {
        if (error.code === 'ENOENT') {
            console.error(`Config file not found: ${configPath}`);
        } else {
            console.error(`Error loading config: ${error.message}`);
        }
        return {};
    }
}

/**
 * Process raw user data into User objects
 */
function processUserData(usersData) {
    const users = [];
    
    for (const data of usersData) {
        try {
            const user = new User(
                data.id,
                data.name,
                data.email,
                data.createdAt ? new Date(data.createdAt) : new Date()
            );
            
            if (validateEmail(user.email)) {
                users.push(user);
            } else {
                console.warn(`Invalid email for user ${data.name}: ${data.email}`);
            }
        } catch (error) {
            console.error(`Error processing user data:`, error);
        }
    }
    
    return users;
}

/**
 * Export user data to JSON file
 */
async function exportUsers(users, filePath) {
    try {
        const data = JSON.stringify(users.map(u => u.toJSON()), null, 2);
        await fs.writeFile(filePath, data, 'utf8');
        console.log(`Exported ${users.length} users to ${filePath}`);
    } catch (error) {
        console.error(`Error exporting users: ${error.message}`);
    }
}

/**
 * Main demo function
 */
async function main() {
    const manager = new UserManager();
    
    // Test authentication
    const user = await manager.authenticate('admin', 'secret123');
    if (user) {
        console.log(`Authentication successful for ${user.name}`);
    } else {
        console.log('Authentication failed');
        process.exit(1);
    }
    
    // Add test users
    const testUsers = [
        new User(1, 'Alice Johnson', 'alice@example.com'),
        new User(2, 'Bob Smith', 'bob@company.org'),
        new User(3, 'Charlie Brown', 'charlie@example.com'),
        new User(4, 'Diana Prince', 'diana@example.com'),
    ];
    
    for (const user of testUsers) {
        if (manager.addUser(user)) {
            console.log(`Added user: ${user.name}`);
        } else {
            console.log(`Failed to add user: ${user.name}`);
        }
    }
    
    // Search functionality
    const searchResults = manager.searchUsers('alice');
    console.log(`Search results for 'alice':`, searchResults.map(u => u.name));
    
    // Get statistics
    const stats = manager.getUserStats();
    console.log('User statistics:', JSON.stringify(stats, null, 2));
    
    // Export data
    await exportUsers(manager.getAllUsers(), './users_export.json');
}

// Export functions for testing
export {
    User,
    UserManager,
    hashPassword,
    validateEmail,
    loadConfig,
    processUserData,
    exportUsers
};

// Run if this is the main module
if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}